clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -268,34 +283,42 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
     /**
      * Get a list of batches which have been sitting in the accumulator too long and need to be expired.
      */
-    public List<ProducerBatch> expiredBatches(int requestTimeout, long now) {
+    public List<ProducerBatch> expiredBatches(long now) {
         List<ProducerBatch> expiredBatches = new ArrayList<>();
+
+        // Expiration of inflight batches happens here in the order of createdMs if the final state of the
+        // batch is not known by (batch's creation time + deliveryTimeoutMs).
+        while (!inFlightBatches.isEmpty() &&
+               now >= inFlightBatches.firstEntry().getKey() + deliveryTimeoutMs) {
+            for (ProducerBatch batch: inFlightBatches.pollFirstEntry().getValue()) {
+                // If the final state of the batch is known (success or failed) just skip past it because
+                // the callback has already been invoked. This allows the clean up of the map in bulk right here.
+                if (batch.finalState() == null) {
+                    batch.expire(now);
```

Comment:
Should we still expire the batches when they are expired instead of expiring all the bucket? Having a second granularity bucket does not prevent us from doing that, right?